cpp/src/arrow/compute/exec/tpch_benchmark.cc

```diff
@@ -0,0 +1,123 @@
+      call("multiply", {field_ref("L_EXTENDEDPRICE"), discount_multiplier});
+  Expression charge =
+      call("multiply",
+           {call("cast",
+                 {call("multiply", {field_ref("L_EXTENDEDPRICE"), discount_multiplier})},
+                 compute::CastOptions::Unsafe(decimal(12, 2))),
+            tax_multiplier});
+  Expression discount = field_ref("L_DISCOUNT");
+
+  std::vector<Expression> projection_list = {l_returnflag, l_linestatus, quantity,
+                                             base_price,   disc_price,   charge,
+                                             quantity,     base_price,   discount};
+  std::vector<std::string> project_names = {
+      "l_returnflag", "l_linestatus", "sum_qty",   "sum_base_price", "sum_disc_price",
+      "sum_charge",   "avg_qty",      "avg_price", "avg_disc"};
+  ProjectNodeOptions project_opts(std::move(projection_list), std::move(project_names));
+
+  ScalarAggregateOptions sum_opts = ScalarAggregateOptions::Defaults();
+  CountOptions count_opts(CountOptions::CountMode::ALL);
+  std::vector<arrow::compute::internal::Aggregate> aggs = {
+      {"hash_sum", &sum_opts},  {"hash_sum", &sum_opts},    {"hash_sum", &sum_opts},
+      {"hash_sum", &sum_opts},  {"hash_mean", &sum_opts},   {"hash_mean", &sum_opts},
+      {"hash_mean", &sum_opts}, {"hash_count", &count_opts}};
+
+  std::vector<FieldRef> to_aggregate = {"sum_qty",    "sum_base_price", "sum_disc_price",
+                                        "sum_charge", "avg_qty",        "avg_price",
+                                        "avg_disc",   "sum_qty"};
+
+  std::vector<std::string> names = {"sum_qty",    "sum_base_price", "sum_disc_price",
+                                    "sum_charge", "avg_qty",        "avg_price",
+                                    "avg_disc",   "count_order"};
+
+  std::vector<FieldRef> keys = {"l_returnflag", "l_linestatus"};
+  AggregateNodeOptions agg_opts(aggs, to_aggregate, names, keys);
+
+  SortKey l_returnflag_key("l_returnflag");
+  SortKey l_linestatus_key("l_linestatus");
+  SortOptions sort_opts({l_returnflag_key, l_linestatus_key});
+  OrderBySinkNodeOptions order_by_opts(sort_opts, sink_gen);
+
+  Declaration filter_decl("filter", {Declaration::Input(lineitem)}, filter_opts);
+  Declaration project_decl("project", project_opts);
+  Declaration aggregate_decl("aggregate", agg_opts);
+  Declaration orderby_decl("order_by_sink", order_by_opts);
+
+  Declaration q1 =
+      Declaration::Sequence({filter_decl, project_decl, aggregate_decl, orderby_decl});
+  std::ignore = *q1.AddToPlan(plan.get());
+  return plan;
+}
+
+static void BM_Tpch_Q1(benchmark::State& st) {
+  for (auto _ : st) {
+    st.PauseTiming();
+    AsyncGenerator<util::optional<ExecBatch>> sink_gen;
+    std::shared_ptr<ExecPlan> plan = Plan_Q1(&sink_gen, static_cast<int>(st.range(0)));
+    st.ResumeTiming();
+    auto fut = StartAndCollect(plan.get(), sink_gen);
+    auto res = *fut.MoveResult();
+  }
+}
```

Comment:
I see this benchmark only outputs a timing in nanoseconds. Do we want to output a more meaningful indicator as well (items per second? something else?)?

Reply:
Q1 only returns four rows ever, so items per second might not be super useful, unless you mean input rows. however, Weston is creating a separate tool anyway so it this benchmark will be deleted very soon anyway in favor of that testing tool, so i don't think it's worth it adding more metrics to this one.

Reply:
I'm not sure the benchmarks will be entirely deleted.  The separate tool is for profiling and not benchmarking.  Ideally we can have one "builtin queries" spot where these queries are defined and then we can use them in both the query tester (for profiling) and the benchmarks (e.g. for detecting regressions).

I agree that output items per second is probably meaningless.  In the long term I think we would want something like "(input rows * cost_factor) per second" where `cost_factor` is some kind of empirically calculated estimate based on how long a query "should" take based on the complexity.  E.g. a scan-filter-project-sink would have a relatively low cost factor but a query with joins would have a higher cost factor.

That's somewhat theoretical though.  Until then I think a relatively intuitive and meaningful metric would be "input bytes per second".  It may not be a simple thing to add though since the size of the input is somewhat random and so we'd need to keep some kind of counters while running.

Let's add a follow-up PR to address this more thoroughly.  It can just be "add TPC-H benchmarks to the benchmarks suite".

Reply:
Hmm, can't the metric simply be "queries per second"? Or does it not make sense for TPC-H?

Reply:
It makes sense in the same way the total time makes sense but it is hard to draw any conclusions from such a number.  For example, we can't easily compare it with our parsing & decoding times to know if our query is an order of magnitude slower or faster than, for example, reading a CSV file or reading a parquet file.

If we want, for example, 1GiBps throughout through some system then it would be nice to get a general sense for which parts of the system will be too slow or the most likely bottlenecks.